### DISCUSSION TITLE
Kernel debugging support

### DISCUSSION THREAD

**Discussion by jfernandez** [arighi/virtme-ng#101] [Ideas]
This may already be supported, so I thought I'd ask before I try to figure this out myself. I want to have a Kernel debugging setup with virtme-ng following these instructions: https://www.kernel.org/doc/html/latest/dev-tools/kgdb.html

gdb would run on the host machine, and kgdb in the VM. How would I go about configuring this?

**Comment by arighi** [comment]
Honestly I've never used vng with kgbd, the closest thing that I've used is to generate a memory dump and use drgn to inspect it, something like this: https://github.com/arighi/virtme-ng/blob/ee99d23fccd269e20f4f3e905b05f0a0b9d0cdc3/README.md?plain=1#L301

That is not the same as attaching gdb directly via kgdb, maybe someone else has more experience than me on this and wants to comment...

**Comment by jfernandez** [comment]
@arighi I can take a stab at trying to make this work. If I enable networking in the VM, I should be able to connect to it from my laptop right?

**Comment by arighi** [comment]
@jfernandez sure! To access the VM from your host you can use `vng --net bridge`, in this way the vng instance should get an IP in the same subnet of the host (assuming you have dhcp) and then anything on the subnet can connect directly to the guest (host included of course).

If you have troubles with `--net bridge` check the troubleshooting section:
https://github.com/arighi/virtme-ng?tab=readme-ov-file#troubleshooting

**Comment by jfernandez** [comment]
@arighi I tried enabling bridge mode and ran into several issues not covered by your troubleshooting section. Instead I used the `--net user` mode and modified the `run.py` params to do port forwarding with `hostfwd`. That worked, but kgdb is meant to be accessed via a serial port, not tcp. So you also need to run something like `socat` inside the VM to bridge between tcp and the serial port.

I think the best solution is to simply enable serial port communication between the host and VM. Then socat and libvirt are not requirements  because networking is not needed for debugging. I can hack virtme-ng to do this, but I found already some serial port params being used and it was unclear if you are already setting this up.

**Reply by arighi** [reply]
Thank you so much for looking at this! Yeah, it would be really useful to have an option in virtme-ng to enable the extra params for a serial port that can be used with kgdb. At the moment the serial stuff is used only for stdin/stderr/stdout redirection (via virtconsole), do the extra options clash with that? For example, are you still able to redirect the kernel log to stderr with `vng -v -- uname -r 2>/tmp/kernel.log`?

**Comment by jimc** [comment]
fwiw, Im using gdb to control a virtme-ng session, heres my wrappers
``
GDESC=`git describe`

function vrun__ () {
    vng --verbose --name $GDESC \
	--user root --cwd ../.. \
	-a dynamic_debug.verbose=2 \
	$*
}
alias vrun_1="vrun__ -p 1"
alias vrun_="vrun__ -p 4 "
alias vrun_3="vrun_ -a dynamic_debug.verbose=3"
alias vrun_9="vrun_ --force-9p"

alias vrun_e="vrun_ -e $*"
alias vrun_t="vrun_3 -e tools/testing/selftests/dynamic_debug/dyndbg_selftest.sh"

alias vrun_d="vrun_ -a nokaslr -o=-S -o=-s "
alias vrun_d9="vrun_d --force-9p"

function vrun_a () {
    vrun__ \
	-a drm.debug=0x100 \
	-a i915.dyndbg=class,DRM_UT_CORE,+pfmlt_%class,DRM_UT_KMS,+pfml_%class,DRM_UT_ATOMIC,+pfm \
	$*
}
``
Theres more to the file, but its all old cruft from old virtme days,
the fns above are all I need these days.

**Comment by SPYFF** [comment]
Most of the time I work with the running kernel that comes with the OS. So instead of development, most of the time I just try to figure out how the kernel works, and for that I have to prepare the sources and debuginfo for virtme-ng. This is not a big deal for development, where you probably have the sources compiled with debuginfo. It would be handy to have something that would automatically figure out the debuginfo. Sure, without the appropriate kernel sources, this does not have much use... What do you think, it would be possible to have options like download debuginfo, download kernel source (and prepare it for GDB).

Right now, if I execute `sudo vng -r --debug -a nokaslr` and then GDB, I see this:

```
(gdb) target remote :1234
Remote debugging using :1234
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0xffffffff81c233df in ?? ()
(gdb) bt
#0  0xffffffff81c233df in ?? ()
#1  0xffffffff81c25549 in ?? ()
#2  0xffffffff81c2584c in ?? ()
#3  0xffffffff81128626 in ?? ()
#4  0x0000000000000000 in ?? ()
```

In a dream scenario `--debug` will prepare everything for the user (at least on Debian & Ubuntu, but I'm sure it is possible on all ditro). Maybe `nokaslr` should be appended by default too. I'm glad that there are efforts to make DWARF-less debugging possible around the drgn project (https://github.com/osandov/drgn/issues/176), so in the future the kernel image+source could be enough for debugging. Few ideas:

1. If `-r` and `--debug` given, `virtme-ng` check distro default places for debuginfo
2. If no debuginfo, ask the user about downloading
3. Also ask about download linux-sources
4. If download not possible or the user dont want it, at least try resolve `kallsyms` to have meaningful backtraces
5. Maybe cache sources/debuginfos for the current kernel

What do you think @arighi, these are intrusive changes? I just thinking about improve the experience of the casual experimenter (like myself) who want to explore the operation of the kernel actually running on the machine.

**Reply by arighi** [reply]
@SPYFF I think this is a great idea. Making kernel debugging accessible is one of the primary goal of virtme-ng. So it'd be really cool if vng could be able to automatically fetch all the required debugging information for the distro kernels.

However, these are the potential issues / challenges that I see:
 1. installing debugging symbols may require to perform root actions (like `sudo apt install linux-image-$(uname -r)-dbgsym` on Ubuntu)
 1. cross-distro support can be challenging (each distro has its own way to provide kernel debugging symbol)
 1. maintenance burden (we need proper automated tests across distro to make sure we don't break this feature is something changes in the distro)

About 1 we may decide to ignore the problem and simply assume that it's up to the user to install the proper debugging package (debuginfod in Ubuntu doesn't provide debugging symbols for the kernel, so we need to explicitly install the proper dbgsym package). Assuming the package is installed vng could just provide a simple gdb wrapper that looks in `/usr/lib/debug/boot/vmlinux-$(uname -r)` (on Ubuntu for example) or any other known location to search for the kernel debugging symbols and run the proper gdb command to attach to the running vng instance and automatically load the kernel symbols.

About 2 and 3 maybe we can start small, like only support Ubuntu and Debian and then slowly add support to other distro (hopefully someone else will be interested and can contribute to improve this feature).

I also like the idea to set nokaslr automatically when `--debug` is used. Feel free to send a PR for this. Again, maybe we can just start small, applying tiny little changes like this one and slowly build up a better support to automatically provide kernel debugging symbols.

I'll think more about this and will try to make something for Ubuntu.

**Comment by arighi** [comment]
@SPYFF it's still a draft for now, but what do you think about #123?

**Comment by jeromecst** [comment]
Currently, the `--debug` flag doesn't allow the use of `monitor` command (which is used to interact with [kdb](https://www.kernel.org/doc/html/latest/dev-tools/kgdb.html#running-kdb-commands-from-gdb) inside gdb). By modifying a couple of options, I found a way to make it work:

`vng --disable-microvm -v --qemu-opts="-serial tcp::1234,server,nowait" -a kgdboc=ttyS1 -a kgdbwait -a nokaslr`


For some reason, I can't connect to the VM (through gdb) using the serial port when `microvm` is enabled. Additionally, the VM won't boot with the `--debug` option set, yielding the error: `qemu-system-x86_64: vmcoreinfo: Could not open 'vmcoreinfo': No such file or directory.`

That's the reason behind `--disable-microvm` and also why `--debug` is not used.

(Note: `kgdbwait` is optional)

Example session:

```sh
~/linux $ vng --gdb
kernel version = 6.5.7-virtme
Reading symbols from vmlinux...
Remote debugging using localhost:1234
kgdb_breakpoint () at kernel/debug/debug_core.c:1219
1219		wmb(); /* Sync point after breakpoint */
(gdb) monitor ps
11 idle processes (state -) and 
91 sleeping system daemon (state [ims]) processes suppressed,
use 'ps A' to see all.
Task Addr               Pid   Parent [*] cpu State Thread             Command
0xffff888003658000        1        0  1    6   R  0xffff888003658b80 *swapper/0

0xffff888003658000        1        0  1    6   R  0xffff888003658b80 *swapper/0
(gdb) 
```
